Adding Docker API endpoint to inspect image manifest

nishanttotla opened the pull request:
The goal of this PR is to make it possible for the client to figure out the image manifest, which will allow the client to construct the spec with the full digest (instead of doing it in the daemon like #28173) and platform information (#31348).

The client can make a call to the Engine to retrieve the manifest.

Here are more details about the change:

**- What I did**
Added `/distribution/{name:*}/json` endpoint in the API, that accesses the registry and pulls the manifest. This is required to pull digest and platform information.

**- How I did it**
I added a new `distributionRouter`, and a function called `getDistributionInfo` in `distribution_routes.go` that accesses the registry to get the digest, and pulls the image manifest to get more information. A `DistributionInspect` struct is returned.

**- How to verify it**
Send a `curl` request to the engine, at `http:/distribution/alpine:latest/json`

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->
Adding Docker API endpoint to inspect image manifest

**- A picture of a cute animal (not mandatory but encouraged)**
(\\__/)
(='.'=)
(")_(")

I'm on a weak wifi, so here's a text bunny.

**What I will do after this PR**:
- Pin image by digest at the client (https://github.com/moby/moby/pull/32388)
- Add platform information to Service spec (https://github.com/docker/swarmkit/pull/1981 and more)

nishanttotla commented on api/server/router/image/image_routes.go:
```diff
@@ -342,3 +350,86 @@ func (s *imageRouter) postImagesPrune(ctx context.Context, w http.ResponseWriter
+			manifestInspect.Platforms = append(manifestInspect.Platforms, m.Platform)
+		}
+	} else if schema2mnfst, ok := mnfst.(*schema2.DeserializedManifest); ok {
+		fmt.Println("IMAGEROUTES: GOT A SCHEMA2 MANIFEST", len(schema2mnfst.References()))
```
Not sure how to retrieve platform information from a `schema2.DeserializedManifest`. Working on figuring this out.

nishanttotla commented on api/server/router/image/image_routes.go:
```diff
@@ -342,3 +350,86 @@ func (s *imageRouter) postImagesPrune(ctx context.Context, w http.ResponseWriter
+		manifestInspect.Digest = dscrptr.Digest
+	} else {
+		manifestInspect.Digest = canonicalRef.Digest()
+		// TODO(nishanttotla): need to set distrepo here as well
```
This part is a little unclear based on the `backed.GetRepository` function as it is right now. Maybe it should be possible for it to just take a `reference.Named` object instead of `reference.NamedTagged`, since it doesn't seem to depend on having the latter.

nishanttotla commented:
Ping @aluzzardi @aaronlehmann @tonistiigi @vieux @icecrime @friism

nishanttotla commented:
Bumping up this PR. It works in the current form, and the goal is for the client to be able to use this endpoint to get manifest information to construct an appropriate spec.

vieux commented on api/server/router/image/image.go:
```diff
@@ -44,6 +44,7 @@ func (r *imageRouter) initRoutes() {
 		router.Cancellable(router.NewPostRoute("/images/{name:.*}/push", r.postImagesPush)),
 		router.NewPostRoute("/images/{name:.*}/tag", r.postImagesTag),
 		router.NewPostRoute("/images/prune", r.postImagesPrune),
+		router.NewPostRoute("/images/{name:.*}/manifest", r.postImagesManifest),
```
why is it a post ?

nishanttotla replied:
I made it POST because I assumed we're sending registry auth token for private repos. But since we send it in the header instead of the payload, is GET fine?

vieux replied:
I believe GET is fine

nishanttotla replied:
@vieux changed it to GET.

aluzzardi replied:
+1 on the `GET`

nishanttotla commented:
Architecture information is updated for all types of manifests now.

cc @aaronlehmann

aaronlehmann commented on api/types/registry/registry.go:
```diff
@@ -102,3 +105,14 @@ type SearchResults struct {
+// TODO(nishanttotla): Decide which other fields to add here
+type ManifestInspect struct {
+	// Digest is the content addressable digest for the image on the registry
+	Digest digest.Digest
```
What about returning the top-level manifest itself instead of (or in addition to) the digest? I could imagine this being useful for some things.

We could also return an optional config blob if one is referenced by the top-level manifest.

Then the platform determination could be done on the client side, making this `/manifest` more generic.

It's still not really ideal, because the daemon is deciding what items to fetch from the registry and return to the client. I kind of like the idea of proxying the registry API through the daemon, but that's a pretty big project and probably way out of scope for what we're trying to do here.

That said, I don't feel strongly about any of this, and I could be okay with just returning specific fields such as `Platforms` in the API response if that's what people prefer.

nishanttotla replied:
@aaronlehmann I'm inclined towards making it as general as possible. So returning the top-level manifest (and config blob) sounds like a better idea, especially because then the client can deal with it as it likes.

nishanttotla replied:
@aaronlehmann as an example, what do you think about
```
{
  Digest       digest.Digest,
  Manifest     distribution.Manifest,
  ConfigJSON   []bytes,
}
```
I still think we'd have to separate cases (fat, v2, v1 manifests) to get the right config blob, but then the client could Unmarshal it as desired.

aaronlehmann replied:
Yeah, that's basically what I had in mind. Although maybe `Manifest` should be `[]byte` so that we aren't introducing distribution types into the API, and also so that the digest can be verified. It's ugly to put JSON inside JSON though.

nishanttotla replied:
Do you mean marshaling the manifest into `[]byte` before sending? I have a couple of questions
- Why does the digest need to be verified at the client?
- Why does sending `[]bytes` help with it?

> It's ugly to put JSON inside JSON though

This  applies to `ConfigJSON` as well, right?

aaronlehmann replied:
The manifest is a JSON document. `DeserializedManifest` is just a convenience for working with it in Go code. If you marshal it to JSON, you'll get the same set of bytes back, so the hash won't change (`DeserializedManifest` implements a `MarshalJSON` method that ensures this).

However, I've thought about it some more, and I don't think using `[]byte` is worth the ugliness. Let's do this instead:

```go
{
  Digest       digest.Digest,
  Manifest     *json.RawMessage,
  ConfigJSON   *json.RawMessage,
}
```

`Manifest` would contain the original manifest JSON returned by `(*DeserializedManifest).MarshalJSON`. This is similar to using `[]byte`, but it causes the JSON to be inserted directly instead of as a string. We lose the ability to verify the digest (because of possible whitespace added around the value during the marshalling process), but we avoid the ugliness off putting JSON inside a string inside JSON.

> Why does the digest need to be verified at the client?

I don't think there's an immediate reason to verify the digest at the client, but I was thinking that if we're going to send the manifest to the client, we might as well send the exact manifest.

> Why does sending []bytes help with it?

Because the hash is computed over these bytes.

> > It's ugly to put JSON inside JSON though

> This applies to ConfigJSON as well, right?

Yes.

nishanttotla replied:
@aaronlehmann PTAL, does this look reasonable?

aaronlehmann commented on api/server/router/image/image_routes.go:
```diff
@@ -342,3 +349,106 @@ func (s *imageRouter) postImagesPrune(ctx context.Context, w http.ResponseWriter
+	if !ok {
+		if _, ok := ref.(reference.Digested); ok {
+			// full image ID
+			return nil
```
Presumably this should return an error

aaronlehmann commented on api/server/router/image/image_routes.go:
```diff
@@ -342,3 +349,106 @@ func (s *imageRouter) postImagesPrune(ctx context.Context, w http.ResponseWriter
+		configJSON []byte
+	)
+
+	if fatmnfst, ok := mnfst.(*manifestlist.DeserializedManifestList); ok {
```
non-blocking: it might be slightly more readable to use a type switch here. It's purely a cosmetic suggestion, though.

aaronlehmann commented on api/types/registry/registry.go:
```diff
@@ -102,3 +105,18 @@ type SearchResults struct {
+	// possibly obtained by looking at the fat manifest
+	Platforms []manifestlist.PlatformSpec
+	// Manifest contains the original manifest JSON
+	Manifest *json.RawMessage
```
I guess since we're potentially returning different kinds of manifests in this field (schema1, schema2, manifest list...), we should have another field containing the media type. Depending which kind of manifest is being returned, this field would either be set to `schema1.MediaTypeSignedManifest`, `schema2.MediaTypeManifest`, or `manifestlist.MediaTypeManifestList`.

nishanttotla replied:
Ah, I intended to add this but missed it. Doing it now.

nishanttotla commented:
@aaronlehmann addressed all comments.

aaronlehmann commented on api/types/registry/registry.go:
```diff
@@ -102,3 +104,19 @@ type SearchResults struct {
+	// ManifestType is the type of manifest. It could be one of
+	// manifestlist.MediaTypeManifestList, schema2.MediaTypeManifest,
+	// or schema1.MediaTypeSignedManifest
+	ManifestType string
```
Maybe `ManifestMediaType`?

aaronlehmann commented on api/types/registry/registry.go:
```diff
@@ -102,3 +104,19 @@ type SearchResults struct {
+	// Manifest contains the original manifest JSON
+	Manifest *json.RawMessage
+	// ConfigJSON contains the image config JSON
+	ConfigJSON *json.RawMessage
```
I suggest adding the `omitempty` flag, as this field won't always be present.

aaronlehmann commented on api/server/router/image/image_routes.go:
```diff
@@ -342,3 +349,106 @@ func (s *imageRouter) postImagesPrune(ctx context.Context, w http.ResponseWriter
+	mnfstObj := json.RawMessage(mnfstJSON)
+	configObj := json.RawMessage(configJSON)
+	manifestInspect.Manifest = &mnfstObj
+	manifestInspect.ConfigJSON = &configObj
```
Let's move the lines related to `ConfigJSON` into the `*schema2.DeserializedManifest` case so that `manifestInspect.ConfigJSON` remains `nil` when other types of manifests are returned.

nishanttotla replied:
@aaronlehmann I wasn't sure if there was a `ConfigJSON` we could return for FAT manifests, so I put it here. It doesn't seem to be the case, so I'll move this.

aaronlehmann commented on api/server/router/image/image_routes.go:
```diff
@@ -342,3 +349,106 @@ func (s *imageRouter) postImagesPrune(ctx context.Context, w http.ResponseWriter
+	}
+	// at this point, we have a digest, so we can retrieve the manifest
+
+	mnfstsrvc, _ := distrepo.Manifests(ctx)
```
Check for an error

aaronlehmann commented on api/server/router/image/image_routes.go:
```diff
@@ -342,3 +349,106 @@ func (s *imageRouter) postImagesPrune(ctx context.Context, w http.ResponseWriter
+	// at this point, we have a digest, so we can retrieve the manifest
+
+	mnfstsrvc, _ := distrepo.Manifests(ctx)
+	mnfst, _ := mnfstsrvc.Get(ctx, manifestInspect.Digest)
```
Check for an error

aaronlehmann commented:
Design LGTM

Don't forget to update `swagger.yml` to document the new API endpoint.

nishanttotla commented:
Thanks for your comments @aaronlehmann, I'll figure out how to update Swagger (maybe a separate PR for that?)

Ping @vieux @aluzzardi

vieux commented:
@nishanttotla sorry to say that, but could you update the swagger in the same PR here is a PR you can use as inspiration: https://github.com/docker/docker/pull/32015/files

nishanttotla commented on api/swagger.yaml:
```diff
@@ -4760,6 +4760,56 @@ paths:
+              Digest: "sha256:58e1a1bb75db1b5a24a462dd5e2915277ea06438c3f105138f97eb53149673c4"
+              ManifestMediaType: "application/vnd.docker.distribution.manifest.v2+json"
+              Manifest:
+              ConfigJSON:
```
@vieux I've updated the swagger file, but not sure how to reference the `Manifest` and `ConfigJSON` types, because they're pointers to `json.RawMessage` which is itself `[]byte`.

Apart from that, I think it looks alright.

vieux replied:
@nishanttotla could you use put a sample one ? inlined like `'{"key":"value","key2":"value2"}'`

nishanttotla commented:
@vieux windows test failure seems unrelated.

vieux commented:
LGTM ping @dnephin

dnephin commented:
This is an API change without any client changes, but my understanding is that the client changes are  the important part.

Why not just have the client connect to the registry directly? I believe all the auth is already in the client, right?

Having a pass through endpoint where two fields are just "blobs" of data is not a good interface.

tonistiigi commented:
@dnephin I believe it to handle cases where registry is behind a firewall.

I'm also interested about the cli changes. Without them it is very hard to have an opinion about this type https://github.com/docker/docker/pull/32061/files#diff-73caf4b121449e3cc9df8ab7554b31baR110 . I don't know why we are exporting digest if we already export full manifest, what is config being used etc.

In the future we will probably store the manifest in daemon. This endpoint should work for the local storage as well then. Do you see any possible isses with that?

Also, there are no tests.

nishanttotla commented:
@tonistiigi here are some client side changes that move the digest pinning to the client side: #32388. Those depend on this PR, and explain part of how this inspect type will be used. Another PR will follow up to add platform information to the service spec, and potentially other relevant things in the future.

We could extend the endpoint to work for local storage too, which shouldn't be hard given that it's in the daemon. I can add tests as a follow up.

tonistiigi commented:
@nishanttotla Thanks. #32388 seems to only use the precomputed `Digest` key, so I'm still confused if this endpoint is for returning `tagged->digested` mappings or for viewing manifests and image configs. If latter then why return schema1 manifest that doesn't have any of the same properties. Or in case it returns manifestlist then the image config part seems useless.

If #32388 is not planned for `17.05` and there are no features blocked on this then I don't recommend pushing this to that release. We shouldn't add a new endpoint if there are no users or tests. If we add it we need to support that data type forever, but I'm not sure we are confident on that before we have some code that actually uses it.

nishanttotla commented:
@tonistiigi this change is not only for viewing `tagged->digested` mappings, but also for viewing the manifest and config. For now, we only wish to extract platform information out of the manifest (see https://github.com/docker/swarmkit/pull/1981, #31348) but it could be more things in the future. Platform information is stored differently depending on whether it's a `schema1`, `schema2`, or `fat` manifest, so we need to include all three.

I can't quite add the part to extract platform information at the client yet, because we first need to be done with #32388, followed by merging and vendoring https://github.com/docker/swarmkit/pull/1981. Both of those PRs are ready to go, but depend on this current one being merged first.

I hope that clarifies why we're doing this.

nishanttotla commented:
@aaronlehmann @thaJeztah @vieux I've updated the PR to have `/distribution/{name:.*}/json` as the endpoint. I've also changed back to only returning relevant fields in a `DistributionInspect` struct. PTAL.

(I still need to update `swagger.yml`, but I'll do that once we agree)

aaronlehmann commented on api/swagger.yaml:
```diff
@@ -4802,6 +4802,50 @@ paths:
           type: "string"
           required: true
       tags: ["Image"]
+  /images/{name}/manifest:
```
Path is outdated

aaronlehmann replied:
Sorry, didn't see that this part was still to come.

aaronlehmann commented on api/swagger.yaml:
```diff
@@ -4802,6 +4802,50 @@ paths:
       tags: ["Image"]
+  /images/{name}/manifest:
+    get:
+      summary: "Get image manifest"
```
It no longer returns the manifest

nishanttotla replied:
@aaronlehmann yeah, I still need to update `swagger.yml`, but was holding off until we agreed on the endpoint and return values.

nishanttotla commented:
Updated `swagger.yml`

aaronlehmann commented on api/server/router/distribution/backend.go:
```diff
@@ -0,0 +1,18 @@
+	distributionBackend
+}
+
+type distributionBackend interface {
```
Why is this private interface necessary?

nishanttotla replied:
It's not necessary, but looking at other backends it seemed like convention to do this. I'm removing the private interface.

aaronlehmann commented on api/server/router/distribution/distribution_routes.go:
```diff
@@ -0,0 +1,121 @@
+		distributionInspect.Digest = dscrptr.Digest
+	} else {
+		distributionInspect.Digest = canonicalRef.Digest()
+		distrepo, _, err = s.backend.GetRepository(ctx, canonicalRef, config)
```
The `GetRepository` calls could be deduplicated if it is called above this `if/else` instead.

nishanttotla replied:
@aaronlehmann fixed, thanks.

aaronlehmann commented on api/types/registry/registry.go:
```diff
@@ -102,3 +105,15 @@ type SearchResults struct {
+	// Digest is the content addressable digest for the image on the registry
+	Digest digest.Digest
+	// ID is the local image digest
+	ID digest.Digest
```
It probably makes sense for `ID` to have the omitempty tag set, since it is not always filled in.

stevvooe replied:
What is `ID` used for? If we can avoid sending it, we can mitigate some of the issues above.

nishanttotla replied:
The `ID` isn't useful immediately, but we have discussed supporting local images on Swarm workers using image IDs instead of pulling them by tag/digest. You're right that `ID` doesn't make sense for manifest lists, so this is limited to only `schema2` images. Does adding an `omitempty` tag set as @aaronlehmann suggested avoid that issue?

stevvooe replied:
> Does adding an omitempty tag set as @aaronlehmann suggested avoid that issue?

No, this breaks the fundamental relationship. Each image could have multiple IDs. You'll have to model is as a collection of ids, but then you are basically doing what the manifest format is doing.

nishanttotla replied:
Based on this, I'm inclined to remove the `ID` field altogether, particularly because we aren't planning to use it right away. We can add it when we actually need it.

stevvooe commented on api/swagger.yaml:
```diff
@@ -8263,3 +8263,62 @@ paths:
           format: "int64"
           required: true
       tags: ["Secret"]
+  /distribution/{name}/json:
```
I would advise caution in placing the name directly in the path. While we do this for the registry, ensuring that all cases are captured is non-trivial, unless it is properly escaped.

thaJeztah replied:
Yes, that issue came up during a meeting earlier this week. Especially colons `:` in URL's can be troublesome (using a `?something=` query-string is better)

it's consistent with all endpoints we have now though, so not sure we should make the exception here (open to input on that one)

stevvooe replied:
As long as it is escaped, it is fine.

thaJeztah replied:
Yup, and that's one thing the API currently doesn't do 😞 - think there was an attempt once to change it, but it would break existing clients

stevvooe commented on api/types/registry/registry.go:
```diff
@@ -102,3 +105,15 @@ type SearchResults struct {
+
+// DistributionInspect describes the result obtained from contacting the
+// registry to retrieve image metadata
+type DistributionInspect struct {
```
This data structure is making the assumption that a name may map to single a image id, which is untrue. For manifest lists, the name may map to multiple image ids (which are effectively the config digest).

nishanttotla replied:
I replied to this in the above thread, that the `ID` only makes sense for `schema2` manifests.

stevvooe replied:
> I replied to this in the above thread, that the ID only makes sense for schema2 manifests.

This is a somewhat narrow assumption. Every actual runnable image can result in an image id, which is the digest of the image config. I wouldn't recommend exposing ID at all, and just perform a digest resolution.

This functionality is represented with [`Resolver.Resolve`](https://github.com/containerd/containerd/blob/master/remotes/resolver.go#L11) in containerd. It encompasses the action taking a name and resolving to an object with a canonical name. This is an identical action here.

stevvooe commented:
How will this endpoint handle OCI and other formats?

nishanttotla commented:
@stevvooe for OCI, is this the manifest format: https://github.com/opencontainers/image-tools/blob/master/image/manifest.go?

If not, could you point me to it? It doesn't seem like `github.com/docker/distribution` handles it right now.

stevvooe commented:
> @stevvooe for OCI, is this the manifest format: https://github.com/opencontainers/image-tools/blob/master/image/manifest.go?

That is correct.

The formats are very close to what is handled by `github.com/docker/distribution`.

nishanttotla commented:
It does look very close, but am I correct in understanding that we don't handing OCI manifests in Docker yet? I can add it to this PR, but also looks like that would be better addressed in a separate PR?

Currently, we're only checking the `Manifest` interface in `github.com/docker/distribution`: https://github.com/docker/distribution/blob/master/manifests.go#L13 and it doesn't seem like the OCI manifest quite fits this interface.

stevvooe commented:
> It does look very close, but am I correct in understanding that we don't handing OCI manifests in Docker yet? I can add it to this PR, but also looks like that would be better addressed in a separate PR?

That is correct, but we need to ensure we don't have a model that prevents us from doing so in the future. The `Resolver.Resolve` approach makes this very simple.

> Currently, we're only checking the Manifest interface in github.com/docker/distribution: https://github.com/docker/distribution/blob/master/manifests.go#L13 and it doesn't seem like the OCI manifest quite fits this interface.

It can.

nishanttotla committed:
Change GetRepository to take Named arguments

Signed-off-by: Nishant Totla <nishanttotla@gmail.com>

nishanttotla committed:
Adding /distribution/{name}/json endpoint to contact registry

Signed-off-by: Nishant Totla <nishanttotla@gmail.com>

vieux commented:
the endpoint LGTM @stevvooe @aaronlehmann any comments ?

aaronlehmann commented:
LGTM

stevvooe commented:
LGTM

My only concern here is that the data type returned here is functionality equivalent to a manifestlist/oci-index. I think we can probably just move to the oci-index at some point and use the `Accept` header to dispatch to this datatype.

stevvooe commented on api/types/registry/registry.go:
```diff
@@ -102,3 +105,13 @@ type SearchResults struct {
+// registry to retrieve image metadata
+type DistributionInspect struct {
+	// Digest is the content addressable digest for the image on the registry
+	Digest digest.Digest
```
Actually, this should be a `Descriptor`.

nishanttotla replied:
@stevvooe do you mean rename the object to `Descriptor` or to use the `Descriptor` struct in blobs?

nishanttotla replied:
@stevvooe do you mean rename the object to `Descriptor` or to use the `Descriptor` struct in blobs?

stevvooe replied:
Like, this should provide the mediatype and size, similar to how we reference all image components.

nishanttotla replied:
@stevvooe starting from a `distribution.Repository` object, it isn't clear how to get the `Descriptor` object cleanly in all cases. When the image has a tag (eg. `latest`), then the `Get(ctx, tag)` function for `TagService` can be used to get the `Descriptor`.

For images that only come in with a digest, the way seems to be calling `Blobs(ctx).Stat(ctx, digest)`, but that is causing an error. I wonder if it's worth adding extra information here that we don't intend to use right now.

cc @aaronlehmann

nishanttotla replied:
@stevvooe updated to return `Descriptor`

nishanttotla committed:
/distribution/{name}/json returns full Descriptor object

Signed-off-by: Nishant Totla <nishanttotla@gmail.com>

stevvooe commented:
LGTM

aaronlehmann commented:
```
23:59:35 The swagger spec at "api/swagger.yaml" is invalid against swagger specification 2.0. see errors :
23:59:35 - must validate one and only one schema (oneOf)
23:59:35 - must validate one and only one schema (oneOf)
23:59:35 - must validate at least one schema (anyOf)
23:59:35 - paths./distribution/{name}/json.get.responses.200.schema.properties.Descriptor.properties.Size.type in body must be of type array: "string"
23:59:35 - must validate one and only one schema (oneOf)
23:59:35 - must validate one and only one schema (oneOf)
23:59:35 - must validate at least one schema (anyOf)
23:59:35 - paths./distribution/{name}/json.get.responses.200.schema.properties.Descriptor.properties.Size.type in body must be of type array: "string"
23:59:35 - must validate one and only one schema (oneOf)
23:59:35 - must validate one and only one schema (oneOf)
23:59:35 - must validate at least one schema (anyOf)
23:59:35 - paths./distribution/{name}/json.get.responses.200.schema.properties.Descriptor.properties.Size.type in body must be of type array: "string"
23:59:35 - must validate one and only one schema (oneOf)
23:59:35 - must validate one and only one schema (oneOf)
23:59:35 - must validate at least one schema (anyOf)
23:59:35 - paths./distribution/{name}/json.get.responses.200.schema.properties.Descriptor.properties.Size.type in body must be of type array: "string"
```

aaronlehmann commented on api/server/router/distribution/distribution_routes.go:
```diff
@@ -0,0 +1,117 @@
+		}
+		distributionInspect.Descriptor = dscrptr
+	} else {
+		dscrptr, err := blobsrvc.Stat(ctx, canonicalRef.Digest())
```
I don't think  `blobsrvc.Stat` works for manifests. Is it better to manually assemble a descriptor using the content of the manifest? (cc @stevvooe)

nishanttotla replied:
@aaronlehmann Would it be okay to fix this as a follow-up, so that items dependent on this PR can be unblocked?

stevvooe replied:
@aaronlehmann Yes, although we should fix this to allow lookup of manifests as a blob.

aaronlehmann replied:
@nishanttotla: I believe we have enough time to fix this before the freeze. I would rather deal with it right now than merge code with known issues.

nishanttotla replied:
@aaronlehmann I'm not sure how to fix it, but I'm happy to work on it. Also, I didn't intend that we leave this as is, just that we could do it as a bug fix (also before the freeze) in parallel with the other changes.

aaronlehmann replied:
The components of `Descriptor` are the digest, size, and media type. We already have the digest here (`canonicalRef.Digest()`). `mnfst.Payload()` returns both the media type and the payload, and with `len(payload)` we can get the size. So just stick those inside a `Descriptor` struct and we're good to go.

nishanttotla replied:
That sounds good to me. I will still leave a comment in there nothing that we should use `Stat` when we're able to.

nishanttotla commented:
@aaronlehmann sorry about that, fixed it.

nishanttotla commented:
@aaronlehmann PTAL.

aaronlehmann commented:
Just one comment... it may be a good idea to override the descriptor's media type always, not just in the `descriptorFromPayload` case, because we know the one the registry give us is going to be misleading. Other than that, it looks good.

nishanttotla commented:
@aaronlehmann sure, updating the PR. If we're going to call `Payload()` always, then `descriptorFromPayload` isn't really needed, and we can update size if it's 0.

aaronlehmann reviewed (approved):
Possible cleanup: `dscrptr` is no longer necessary because `distributionInspect.Descriptor` can be set directly. Not important though.

LGTM

nishanttotla commented:
@aaronlehmann thanks, I cleaned that up.

thaJeztah commented on api/swagger.yaml:
```diff
@@ -8274,3 +8274,71 @@ paths:
+                    Features:
+                      type: "array"
+                      items:
+                        type: "string"
```
Perhaps we should add an example response here, because the current output doesn't provide a lot of information to the reader :smile:
<img width="546" alt="screen shot 2017-05-10 at 19 13 22" src="https://cloud.githubusercontent.com/assets/1804568/25911743/45b22992-35b5-11e7-8335-5e565cfe9bc8.png">

nishanttotla replied:
@thaJeztah updated. PTAL.

<img width="820" alt="screen shot 2017-05-10 at 11 02 24 am" src="https://cloud.githubusercontent.com/assets/1872537/25913611/3983bad6-3570-11e7-9e5c-a1af1312c768.png">

nishanttotla replied:
Looks like empty array definitions are still causing an error, I'm fixing those.

thaJeztah replied:
Can we come up with some sample URL's and/or fill the OSVersion / Variants? Or do you think it's ok to keep them empty?

nishanttotla replied:
@thaJeztah In the examples I saw, these weren't filled, so I think it's okay. I can try to search for better examples later.

thaJeztah commented on api/swagger.yaml:
```diff
@@ -8274,3 +8274,71 @@ paths:
+                    type: "string"
+                  Size:
+                    type: "integer"
+                    format: "int64"
```
Wondered if this should be an `uint64`, but I see it's coming from an existing type

nishanttotla committed:
Adding example to /distribution/{name}/json endpoint swagger spec

Signed-off-by: Nishant Totla <nishanttotla@gmail.com>

nishanttotla commented:
@thaJeztah PTAL.

thaJeztah commented:
I think We're almost there; left one comment on the example, also (sorry, forgot to mention); this needs an entry in the API history; https://github.com/moby/moby/blob/master/docs/api/version-history.md#v130-api-changes

nishanttotla commented:
@thaJeztah updated API history here: #33147

mdlinville reviewed (approved):
Docs LGTM

thaJeztah reviewed (approved):
LGTM, thanks!
